docs/proposal/20211207-delegated-storage.md

```diff
@@ -0,0 +1,1186 @@
+    capact-outputTypeInstances:
+      - name: helm-release
+        from: helm-release
+        backend: helm-storage # new property -> alias defined in the `requires` section
+    ```
+
+1. The automatically injected TypeInstance upload step, receives the following payload:
+
+    ```yaml
+    typeInstances:
+    - alias: helm-release
+      attributes: []
+      createdBy: default/act-l49vh-30c7a078-6a77-475c-94dd-7466f56447ce
+      typeRef:
+        path: cap.type.helm.chart.release
+        revision: 0.1.0
+      value: null 
+      backend:
+        # Fields which are a part of TypeInstance
+        id: 3ef2e4ac-9070-4093-a3ce-142139fd4a16 # helm-release backend - resolved UUID based on the injected TypeInstance
+
+        # Fields which are a part of TypeInstanceResourceVersion (can be changed later via TypeInstance Update):
+        additionalParameters:
+          name: release-name
+          namespace: release-namespace
+    usesRelations: # automatically create relation between TypeInstance using a given backend
+    - from: helm-release
+      to: 3ef2e4ac-9070-4093-a3ce-142139fd4a16
+    ```
+
+1. Hub receives the following GraphQL mutation based on the payload fields from point above:
+
+    ```graphql
+    mutation CreateTypeInstances {
+      createTypeInstances(
+        in: {
+          # ... payload
+        }
+      ) {
+        id
+        alias
+      }
+    }
+    ```
+
+1. Based on the `backend` data:
+
+    1. Hub resolves details about the service (TypeInstance details)
+    1. Hub validates whether the storage backend accepts TypeInstance value (`acceptValue` property). If not, and the value has been provided, it returns error.
+    1. Hub calls the registered storage backend service `onCreate` hook:
+
+        ```proto
+        message TypeInstanceData {
+          string id = 1;
+          google.protobuf.Any value = 2;
+        }
+
+        message OnCreateRequest {
+          TypeInstanceData typeinstance = 1;
+          google.protobuf.Any additional_parameters = 2;
+        }
+
+        message OnCreateResponse {
+          google.protobuf.Any additional_parameters = 1;
+        }
+
+        service SearchService {
+          rpc OnCreate(OnCreateRequest) returns (OnCreateResponse);
+        }
+        ```
+
+        This hook can mutate `additionalParameters`.
+
+    1. Hub validates `additionalParameters` against JSON schema saved in the storage backend TypeInstance.
+    1. Saves TypeInstance metadata in the core Hub storage backend, which contains all metadata of the TypeInstances and  theirs relations.
+
+      ``` yaml
+      id: 5d925774-55d3-4c74-8ecb-c984ea7ef636
+      typeRef:
+        path: cap.core.type.hub.storage.redis
+        revision: 0.1.0
+      latestResourceVersion:
+        resourceVersion: 1
+        backend:
+          additionalParameters: # additional parameters that might be modified via the service handling `onCreate` hook
```

Comment:
> IMO the generated ID needs to be saved in a given backend as we will use it to get the data. 

Could be, but that's not the only option. I wanted to avoid all backends have a dedicated different storage to store the `TypeInstanceID <-> Resource` mapping.

>  In general, this should be avoided, as we don’t want to store any data internally

Why? I would allow the backends to store some additional params to simplify them. And yes, the additional params will be stored in a plain text. There shouldn't be any credentials. If you have better idea how to name this field, please suggest different one 🙂

Reply:
>Why? 

I know that it complicates the Helm storage, but also straights out the responsibility between those components.

That's my personal feeling—as on this backend, we will have more responsibility. In case of any downtime (wrong update, wrong restore etc) in simplified case we will lose only relations, but if we will take that responsibility back then we can lose a data that is required  to locate TypeInstance value back.

Currently, we will write those external storage so yeah, it's a plus for us, but in the future when we will start supporting 3rd party storages -  it could become hard. What if someone will try to save there a lot of data? I know that we can say that it's not our problem but probably it will be 🗡️  😄 

----
if we want to have it for now, maybe:
- **context** - to say that it stores only information that a backend needs to  know to get the data
- **metadata** 
- **lookupInformation**
- **referenceData**/**referenceInfo**


To somehow, say that it's used only to store the information which is used to locate and get a target data.

The **additionalParameters**  for me may contain quite a lot of data as we already have such “idea” (same naming) for input to an action where we put all (impl specific data). Maybe it's only my

Reply:
As discussed f2f - I will rename the field to `context`